### ISSUE TITLE
[Bug] Breaks Shell-mode in Emacs 29 master

### ISSUE THREAD

**Issue by butter-radish** [trevorpogue/topspace#21]
### Describe the bug
Initialising Emacs with `global-topspace-mode` breaks Shell-mode in the master branch of Emacs 29.

When you spawn Shell buffers with Global-TopSpace on, the shell prompt fails to get displayed in buffer. If you try to execute any commands, Emacs freezes up and you have to kill it.

I've narrowed TopSpace down as the culprit for this as commenting it out in my init fixes the issue.

### To reproduce
1.  Turn on `global-topspace-mode` in your init.el (I use use-package, and the code I use is:)
    
        (use-package topspace
            :ensure t
            :config (global-topspace-mode 1)
            )

2.  Launch Emacs

3.  Spawn a Shell buffer (<kbd>M-x shell RET</kbd>)
4.  Type anything and press Return

5.  Freeze up

### Expected behavior
Shell-mode (and Emacs in-turn) to not freeze up. And spawn with the shell prompt.

### TopSpace version information

Version: 20220513.1925
Commit: dd7d35b52dda7bf503b8d4c23d97150152eebc41
	 
### Emacs version

GNU Emacs 29.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.33, cairo version 1.17.6) of 2022-05-22

**Comment by trevorpogue** [comment]
Thanks for finding this issue. I tried this in Emacs 28.1 and it didn't freeze. I'm trying to build Emacs 29 on my machine now to test it but I'm getting some build errors. In the meantime, you could try disabling topspace in `shell-mode` with the following:
```elisp
(defun my-topspace-active () (not (eq major-mode 'shell-mode)))
(setq topspace-active #'my-topspace-active)
```

**Comment by butter-radish** [comment]
Thank you, that code snippet takes care of the bug without having to ditch TopSpace completely.

I've noticed more things playing around with initialising Global-TopSpace on the master branch—I can make my shell prompt *appear* with it.

If I may first reiterate the matter of the prompt 'not appearing' in Shell-mode with TopSpace turned on, which I don't think I described that well in my first comment—what happens is that when I launch a Shell buffer: I don't see my shell prompt, but the cursor appears (and blinks as normal) and I can type and move it around freely at this time. Once I press Return at this point (on a command or gibberish) Emacs freezes up; I get no feedback from my shell, I get no feedback from Emacs (and I can't do anything in the app, as you can imagine).

Now, I've just found that I can 'make the prompt appear', so as to speak, by switching to-and-from the minibuffer (<kbd>M-x</kbd>, `save-buffers-kill-emacs`). Strangely, just switching back to Shell from the minibuffer isn't always enough to make the prompt show itself, some of the times I needed to move my cursor around in Shell-mode (spamming <kbd>C-b</kbd>, <kbd>C-f</kbd>, and <kbd>C-p</kbd>) after closing the minibuffer for the prompt to come back.

However, the things is, making the shell prompt 'appear' like this with the switch-to-minibuffer manoeuvre doesn't fix the issue. Shell-mode proceeds to still freeze up once I try to execute anything, with the prompt superficially visible this time.

Conversely, not loading TopSpace in Shell-mode makes the shell prompt appear normally as the buffer opens, and commands and all run normally.

Additionally, not sure, if this is helpful but the `save-buffers-kill-emacs` dialogue asks me to confirm that I wanted to close the shell process running in Shell-mode—so, it appears Emacs Shell-mode is running the shell; the modeline does say "(Shell:run)" throughout btw.

To gave a timeline: the bug appeared for me on the latest Emacs 29 compiled from master yesterday. I'd last compiled Emacs 29 at the end of March, and I wanna say I installed TopSpace for the first time shortly after that (after seeing it on r/Emacs and Sacha Chua's blog). It has worked without hitches in Shell for me on Emacs 29 before this.

**Comment by trevorpogue** [comment]
I was able to finally build Emacs 29 today after upgrading my linux distro, sorry for the delay. 

I followed the steps to reproduce the bug in your first comment here, but everything worked fine for me with no sign of the bug. Have you tried this with the default values for the `topspace` defcustom variables? If so I guess all I can suggest is to check again if there is another package causing the issue.

I used the most recent `topspace` version from fa67b67e1ae41adb44de0e5180a6cab922da6bb0 and the Emacs 29 build is from an Emacs repo cloned a couple of months ago with no extra options provided to the `./configure` command when building. I also tried a new Emacs clone but was getting a startup bug that I didn't think was worth debugging considering your issue is also from a couple of months ago.
